File: dashboard/src/lib/physical.ts
import type { PhysicalAssessment, TelemetrySample } from '../types'

const CRITICALITY: Record<string, number> = {
  smart_lock: 1,
  smart_camera: 0.88,
  smart_thermostat: 0.84,
  smart_plug: 0.76,
}

const HAZARDS: Record<string, { hazard: string; consequence: string; multiplier: number }> = {
  'smart_lock:unauthorized_commands': {
    hazard: 'forced_entry',
    consequence: 'A hostile session can unlock the door and admit an intruder.',
    multiplier: 1.15,
  },
  'smart_lock:firmware_modification': {
    hazard: 'lock_hijack',
    consequence: 'Unsigned firmware can silently disable locking or ignore owner PINs.',
    multiplier: 1.12,
  },
  'smart_camera:abnormal_network_traffic': {
    hazard: 'privacy_exfiltration',
    consequence: 'Live video from the entryway may be leaving the home.',
    multiplier: 1.08,
  },
  'smart_camera:suspicious_ip_connections': {
    hazard: 'surveillance_hijack',
    consequence: 'An attacker may be watching occupants and casing the property.',
    multiplier: 1.1,
  },
  'smart_thermostat:unauthorized_commands': {
    hazard: 'climate_extremes',
    consequence: 'HVAC can be driven to freeze pipes or overheat occupied rooms.',
    multiplier: 1.1,
  },
  'smart_plug:abnormal_power_usage': {
    hazard: 'electrical_fire',
    consequence: 'Sustained overload on a lamp circuit can ignite nearby furnishings.',
    multiplier: 1.18,
  },
  'smart_lock:multi_stage_compromise': {
    hazard: 'forced_entry',
    consequence: 'A coordinated compromise of the lock can open the home to an intruder.',
    multiplier: 1.2,
  },
}

const GENERIC: Record<string, { hazard: string; consequence: string; multiplier: number }> = {
  multi_stage_compromise: {
    hazard: 'coordinated_physical_compromise',
    consequence: 'Multiple hostile behaviors at once raise the chance of a real-world incident.',
    multiplier: 1.2,
  },
  unauthorized_commands: {
    hazard: 'hostile_actuation',
    consequence: 'Physical actuators may fire without the homeowner’s consent.',
    multiplier: 1.06,
  },
  firmware_modification: {
    hazard: 'untrusted_control',
    consequence: 'The device can no longer be trusted to fail safe.',
    multiplier: 1.08,
  },
  abnormal_power_usage: {
    hazard: 'energy_hazard',
    consequence: 'Power anomalies often precede overheating, mining, or stuck actuators.',
    multiplier: 1.02,
  },
  suspicious_ip_connections: {
    hazard: 'external_command_channel',
    consequence: 'An untrusted host can reach a device that moves or observes the home.',
    multiplier: 1,
  },
  abnormal_network_traffic: {
    hazard: 'data_and_control_leak',
    consequence: 'Unexpected egress can carry sensor data or accept remote control.',
    multiplier: 0.95,
  },
}

const SAFE_MODE: Record<string, string> = {
  smart_lock: 'Force the deadbolt locked, disable remote unlock, and shorten auto-relock.',
  smart_camera: 'Close the privacy shutter, stop recording, and cut the live stream.',
  smart_thermostat: 'Hold an eco setpoint, idle HVAC, and lock climate controls.',
  smart_plug: 'Open the relay, cap power draw, and freeze the schedule.',
}

function sensorSignals(sample: TelemetrySample, threat: string): { boost: number; notes: string[] } {
  const notes: string[] = []
  let boost = 0
  const sensors = sample.sensors
  if (sample.device_type === 'smart_lock') {
    if (sensors.locked === false) {
      boost += 22
      notes.push('The latch is currently unlocked.')
    }
    if (sensors.door_ajar) {
      boost += 16
      notes.push(
        sensors.locked === true
          ? 'The door sensor still reports ajar; locking the latch did not close the leaf.'
          : 'The door is ajar, so a remote unlock becomes physical entry.',
      )
    }
    if (Number(sensors.failed_pin_attempts ?? 0) >= 3) {
      boost += 10
      notes.push('Repeated PIN failures look like a physical bypass attempt.')
    }
  } else if (sample.device_type === 'smart_camera' && threat !== 'benign') {
    if (sensors.privacy_shutter_open && sensors.recording) {
      boost += 8
      notes.push('The shutter is open and the camera is recording.')
    }
    if (Number(sensors.motion_probability ?? 0) >= 0.4) {
      boost += 10
      notes.push('Motion at the entryway while the camera is hostile raises stalking risk.')
    }
  } else if (sample.device_type === 'smart_thermostat') {
    const setpoint = Number(sensors.setpoint_c ?? 21)
    if (setpoint >= 30 || setpoint <= 12) {
      boost += 20
      notes.push(`Setpoint is ${setpoint.toFixed(1)}°C, outside a safe occupied range.`)
    }
  }
  return { boost, notes }
}

export function assessPhysical(sample: TelemetrySample, threat: string, cyberRisk: number): PhysicalAssessment {
  const { boost, notes } = sensorSignals(sample, threat)

  if (threat === 'benign') {
    if (sample.device_type === 'smart_camera' || boost < 8) {
      return {
        score: Number(Math.min(12, cyberRisk * 0.15).toFixed(1)),
        hazard: 'none',
        severity: 'none',
        consequences: ['No material physical hazard predicted from this tick.'],
        safe_mode: 'Device remains in its normal operating posture.',
      }
    }
    return {
      score: Number(Math.min(55, 20 + boost).toFixed(1)),
      hazard: 'unsafe_physical_state',
      severity: 20 + boost >= 55 ? 'high' : 20 + boost >= 30 ? 'moderate' : 'low',
      consequences: notes.length ? notes : ['Sensors show an unsafe posture without a classified cyber attack.'],
      safe_mode: SAFE_MODE[sample.device_type] ?? 'Enter failsafe and disable remote admin.',
    }
  }

  const profile =
    HAZARDS[`${sample.device_type}:${threat}`] ??
    GENERIC[threat] ?? {
      hazard: 'device_misuse',
      consequence: 'Anomalous control of a cyber-physical device can affect the home.',
      multiplier: 0.8,
    }

  const score = Math.min(100, cyberRisk * (CRITICALITY[sample.device_type] ?? 0.6) * profile.multiplier + boost)
  const severity = score >= 80 ? 'critical' : score >= 55 ? 'high' : score >= 30 ? 'moderate' : score >= 15 ? 'low' : 'none'
  return {
    score: Number(score.toFixed(1)),
    hazard: score >= 20 ? profile.hazard : 'watch',
    severity,
    consequences: [profile.consequence, ...notes].slice(0, 4),
    safe_mode: SAFE_MODE[sample.device_type] ?? 'Enter failsafe and disable remote admin.',
  }
}

File: security_engine/scoring/physical.py
from __future__ import annotations

from dataclasses import dataclass

from iot_simulator.models.telemetry import TelemetrySample
from security_engine.models.findings import PhysicalAssessment, ThreatType
from security_engine.response.effects import SAFE_MODE_ACTIONS, infer_device_kind


@dataclass(frozen=True)
class HazardProfile:
    hazard: str
    consequence: str
    multiplier: float


DEVICE_CRITICALITY: dict[str, float] = {
    "smart_lock": 1.0,
    "smart_camera": 0.88,
    "smart_thermostat": 0.84,
    "smart_plug": 0.76,
    "iot_device": 0.6,
}

THREAT_HAZARDS: dict[tuple[str, str], HazardProfile] = {
    ("smart_lock", ThreatType.UNAUTHORIZED_COMMANDS.value): HazardProfile(
        "forced_entry",
        "A hostile session can unlock the door and admit an intruder.",
        1.15,
    ),
    ("smart_lock", ThreatType.FIRMWARE_MODIFICATION.value): HazardProfile(
        "lock_hijack",
        "Unsigned firmware can silently disable locking or ignore owner PINs.",
        1.12,
    ),
    ("smart_lock", ThreatType.SUSPICIOUS_IP_CONNECTIONS.value): HazardProfile(
        "remote_entry_channel",
        "An untrusted peer can reach the lock control plane from outside the home.",
        1.05,
    ),
    ("smart_camera", ThreatType.ABNORMAL_NETWORK_TRAFFIC.value): HazardProfile(
        "privacy_exfiltration",
        "Live video from the entryway may be leaving the home.",
        1.08,
    ),
    ("smart_camera", ThreatType.SUSPICIOUS_IP_CONNECTIONS.value): HazardProfile(
        "surveillance_hijack",
        "An attacker may be watching occupants and casing the property.",
        1.1,
    ),
    ("smart_camera", ThreatType.FIRMWARE_MODIFICATION.value): HazardProfile(
        "covert_recording",
        "A tampered camera can keep recording after the shutter appears closed.",
        1.06,
    ),
    ("smart_thermostat", ThreatType.UNAUTHORIZED_COMMANDS.value): HazardProfile(
        "climate_extremes",
        "HVAC can be driven to freeze pipes or overheat occupied rooms.",
        1.1,
    ),
    ("smart_thermostat", ThreatType.ABNORMAL_POWER_USAGE.value): HazardProfile(
        "hvac_stress",
        "Abnormal draw can damage HVAC equipment and create fire load.",
        1.0,
    ),
    ("smart_plug", ThreatType.ABNORMAL_POWER_USAGE.value): HazardProfile(
        "electrical_fire",
        "Sustained overload on a lamp circuit can ignite nearby furnishings.",
        1.18,
    ),
    ("smart_plug", ThreatType.UNAUTHORIZED_COMMANDS.value): HazardProfile(
        "unsafe_energize",
        "A hijacked relay can energize a load when nobody is home.",
        1.04,
    ),
    ("smart_lock", ThreatType.MULTI_STAGE_COMPROMISE.value): HazardProfile(
        "forced_entry",
        "A coordinated compromise of the lock can open the home to an intruder.",
        1.2,
    ),
    ("smart_camera", ThreatType.MULTI_STAGE_COMPROMISE.value): HazardProfile(
        "surveillance_hijack",
        "A coordinated camera compromise can watch the home and leak the feed.",
        1.16,
    ),
    ("smart_thermostat", ThreatType.MULTI_STAGE_COMPROMISE.value): HazardProfile(
        "climate_extremes",
        "A coordinated HVAC compromise can freeze pipes or overheat occupied rooms.",
        1.14,
    ),
    ("smart_plug", ThreatType.MULTI_STAGE_COMPROMISE.value): HazardProfile(
        "electrical_fire",
        "A coordinated plug compromise can overload a circuit and ignite furnishings.",
        1.18,
    ),
}

GENERIC_HAZARDS: dict[str, HazardProfile] = {
    ThreatType.MULTI_STAGE_COMPROMISE.value: HazardProfile(
        "coordinated_physical_compromise",
        "Multiple hostile behaviors at once raise the chance of a real-world incident.",
        1.2,
    ),
    ThreatType.FIRMWARE_MODIFICATION.value: HazardProfile(
        "untrusted_control",
        "The device can no longer be trusted to fail safe.",
        1.08,
    ),
    ThreatType.UNAUTHORIZED_COMMANDS.value: HazardProfile(
        "hostile_actuation",
        "Physical actuators may fire without the homeowner's consent.",
        1.06,
    ),
    ThreatType.ABNORMAL_NETWORK_TRAFFIC.value: HazardProfile(
        "data_and_control_leak",
        "Unexpected egress can carry sensor data or accept remote control.",
        0.95,
    ),
    ThreatType.SUSPICIOUS_IP_CONNECTIONS.value: HazardProfile(
        "external_command_channel",
        "An untrusted host can reach a device that moves or observes the home.",
        1.0,
    ),
    ThreatType.ABNORMAL_POWER_USAGE.value: HazardProfile(
        "energy_hazard",
        "Power anomalies often precede overheating, mining, or stuck actuators.",
        1.02,
    ),
    ThreatType.UNKNOWN_ANOMALY.value: HazardProfile(
        "uncharacterized_physical_risk",
        "Behavior left the learned envelope; physical side-effects are possible.",
        0.7,
    ),
}


class PhysicalRiskPredictor:
    """Maps cyber findings onto predicted physical harm in the home."""

    def assess(
        self,
        sample: TelemetrySample,
        threat_type: ThreatType | str,
        cyber_risk: float,
    ) -> PhysicalAssessment:
        kind = sample.device_type.value if hasattr(sample.device_type, "value") else str(sample.device_type)
        if kind not in DEVICE_CRITICALITY:
            kind = infer_device_kind(sample.device_id)
        threat = threat_type.value if isinstance(threat_type, ThreatType) else str(threat_type)
        profile = THREAT_HAZARDS.get((kind, threat)) or GENERIC_HAZARDS.get(threat)
        sensor_boost, sensor_notes = self._sensor_signals(kind, sample.sensors, threat)

        if threat in {ThreatType.BENIGN.value, "benign"}:
            # Recording, motion, and an open shutter are expected camera functions.
            if kind == "smart_camera" or sensor_boost < 8:
                return PhysicalAssessment(
                    score=round(min(12.0, cyber_risk * 0.15), 1),
                    hazard="none",
                    severity="none",
                    consequences=["No material physical hazard predicted from this tick."],
                    safe_mode="Device remains in its normal operating posture.",
                )
            return PhysicalAssessment(
                score=round(min(55.0, 20.0 + sensor_boost), 1),
                hazard="unsafe_physical_state",
                severity=_severity(20.0 + sensor_boost),
                consequences=sensor_notes or ["Sensors show an unsafe posture without a classified cyber attack."],
                safe_mode=_safe_mode_copy(kind, SAFE_MODE_ACTIONS.get(kind, SAFE_MODE_ACTIONS["iot_device"])),
            )

        if profile is None:
            profile = HazardProfile(
                "device_misuse",
                "Anomalous control of a cyber-physical device can affect the home.",
                0.8,
            )

        criticality = DEVICE_CRITICALITY.get(kind, 0.6)
        score = min(100.0, cyber_risk * criticality * profile.multiplier + sensor_boost)
        severity = _severity(score)
        consequences = [profile.consequence, *sensor_notes]
        actions = SAFE_MODE_ACTIONS.get(kind, SAFE_MODE_ACTIONS["iot_device"])
        safe_mode = _safe_mode_copy(kind, actions)
        return PhysicalAssessment(
            score=round(score, 1),
            hazard=profile.hazard if score >= 20 else "watch",
            severity=severity,
            consequences=consequences[:4],
            safe_mode=safe_mode,
        )

    def _sensor_signals(self, kind: str, sensors: dict, threat: str = "") -> tuple[float, list[str]]:
        notes: list[str] = []
        boost = 0.0
        benign = threat in {ThreatType.BENIGN.value, "benign", ""}
        if kind == "smart_lock":
            if sensors.get("locked") is False:
                boost += 22
                notes.append("The latch is currently unlocked.")
            if sensors.get("door_ajar"):
                boost += 16
                notes.append(
                    "The door sensor still reports ajar; locking the latch did not close the leaf."
                    if sensors.get("locked") is True
                    else "The door is ajar, so a remote unlock becomes physical entry."
                )
            if int(sensors.get("failed_pin_attempts") or 0) >= 3:
                boost += 10
                notes.append("Repeated PIN failures look like a physical bypass attempt.")
            if sensors.get("tamper_switch"):
                boost += 18
                notes.append("The lock tamper switch is active.")
        elif kind == "smart_camera" and not benign:
            if sensors.get("privacy_shutter_open") and sensors.get("recording"):
                boost += 8
                notes.append("The shutter is open and the camera is recording.")
            if float(sensors.get("motion_probability") or 0) >= 0.4:
                boost += 10
                notes.append("Motion at the entryway while the camera is hostile raises stalking risk.")
        elif kind == "smart_thermostat":
            setpoint = float(sensors.get("setpoint_c") or 21.0)
            if setpoint >= 30 or setpoint <= 12:
                boost += 20
                notes.append(f"Setpoint is {setpoint:.1f}°C, outside a safe occupied range.")
            if sensors.get("hvac_mode") in {"heat", "cool"} and sensors.get("controls_locked") is not True:
                boost += 6
        elif kind == "smart_plug":
            if sensors.get("relay_on") and sensors.get("in_rush_detected"):
                boost += 14
                notes.append("The relay is closed during an in-rush spike.")
            if sensors.get("relay_on"):
                boost += 4
        return boost, notes


def _severity(score: float) -> str:
    if score >= 80:
        return "critical"
    if score >= 55:
        return "high"
    if score >= 30:
        return "moderate"
    if score >= 15:
        return "low"
    return "none"


def _safe_mode_copy(kind: str, actions: list[str]) -> str:
    labels = {
        "smart_lock": "Force the deadbolt locked, disable remote unlock, and shorten auto-relock.",
        "smart_camera": "Close the privacy shutter, stop recording, and cut the live stream.",
        "smart_thermostat": "Hold an eco setpoint, idle HVAC, and lock climate controls.",
        "smart_plug": "Open the relay, cap power draw, and freeze the schedule.",
    }
    fallback = "Enter failsafe and disable remote admin: " + ", ".join(actions[:3])
    return labels.get(kind, fallback)
